### ISSUE TITLE
How to fine-tune pretrained model on downstream dataset?

### ISSUE THREAD

**Issue by zer02her0** [wjq-learning/CBraMod#1]
Hi!

Thanks for your great work on the CBramod model and its implementation! I am currently working on fine-tuning the model and have a few questions regarding the fine-tuning process.

I have carefully read both the paper and readme. As the instructions, I have loaded the pre-trained weights and replaced the reconstruction head with task-specific head. The data preprocessing and hyperparameters are also set according to the paper. However, the results on TUEV are significantly lower than those reported in the paper. Could you please share the fine-tuning code or provide additional guidance?

Thank you so much for your help!

**Comment by wjq-learning** [comment]
Hi!

Thank you for your message and for your interest in the CBraMod model! We appreciate your efforts in following the instructions closely. 

Currently, the project is still being updated, and the fine-tuning code for each dataset is still being organized. We're in the process of finalizing it, so please bear with us a little longer. We’ll make sure to share the fine-tuning code as soon as it's ready.

Thank you for your patience and understanding! Feel free to reach out if you have any other questions in the meantime.

**Comment by zer02her0** [comment]
Thank you for sharing the downstream preprocessing and finetuning code! I'm currently  trying to replicate the BCIC2020-3 results  and have some questions that I hope you can help with: 
1. I used preprocessing_speech.py to preprocess the data and followed the hyperparameters from Labram's paper for finetuning Labram , but the  balanced accuracy is only 0.2226. Could you share your process for finetuning Labram or provide any additional guidance on the data preprocessing steps and finetuning steps?
2. When finetuning cbramod, I obtained a balanced accuracy of 0.496 which is slightly lower than what is reported in the paper. I attach the  training log file for your reference:
[bcic2020_3_seed_3407_log.txt](https://github.com/user-attachments/files/19429380/bcic2020_3_seed_3407_log.txt).

I would greatly appreciate any insights you could offer. Thank you in advance for your help!

**Comment by wjq-learning** [comment]
Thank you for your detailed issue and for sharing the training log file. I’ll address your concerns step by step:

### 1. **Finetuning LaBraM**
For LaBraM, we recommend avoiding the use of the classifier head as described in the original paper and code. The original implementation might rely on the `cls_token` and feed its embedding into the classifier, which could negatively impact performance. Instead, we suggest directly flattening the features output by LaBraM and using the classifier head designed for CBraMod (as shown in the README example). This classifier head can be directly applied to the flattened features from LaBraM with the setting of `return_patch_tokens=True` for LaBraM. This approach has been more effective in our experiments.

### 2. **Balanced Accuracy for BCIC2020-3**
Regarding the balanced accuracy results:
- For LaBraM, the low balanced accuracy (0.2226) may also be influenced by the classifier design mentioned above. Switching to the recommended classifier head should help improve performance.
- For CBraMod, the balanced accuracy of 0.496 is close to the reported results but slightly lower. Have you experimented with other random seeds? We recommend trying additional random seeds, as variability in initialization and data shuffling can sometimes lead to differences in performance. Running multiple trials with different seeds will provide a clearer picture of the model's stability and expected performance.

If further issues persist after these adjustments, please feel free to share additional details or logs, and I’ll be happy to assist further.

**Comment by changjie98** [comment]
I came across your previous discussion regarding finetuning LaBraM and CBraMod on the BCIC2020-3 dataset, where you suggested replacing LaBraM's classifier head with the CBraMod classifier head (using return_patch_tokens=True to flatten the features). I tried this approach but am still encountering performance issues.

Specifically, the model shows severe overfitting during training—training loss decreases rapidly, while validation balanced accuracy remains very low (around 0.24–0.3). This indicates that the current hyperparameter settings might not be well-suited for this configuration.

Could you please share the detailed hyperparameter settings you used for finetuning LaBraM with the CBraMod classifier head on BCIC2020-3? 

Thank you for your time and help. I would be happy to provide any additional details if needed.

**Comment by changjie98** [comment]
Here are my model parameters and results:
Namespace(batch_size=64, epochs=50, update_freq=1, save_ckpt_freq=10, robust_test=None, model='labram_base_patch200_200', qkv_bias=True, rel_pos_bias=True, abs_pos_emb=False, layer_scale_init_value=0.1, input_size=600, drop=0.1, attn_drop_rate=0.1, drop_path=0.1, disable_eval_during_finetuning=False, model_ema=False, model_ema_decay=0.9999, model_ema_force_cpu=False, opt='adamw', opt_eps=1e-08, opt_betas=[0.9, 0.99], clip_grad=0.5, momentum=0.9, weight_decay=0.0001, weight_decay_end=None, lr=1e-05, layer_decay=0.9, warmup_lr=1e-06, min_lr=1e-06, warmup_epochs=5, warmup_steps=-1, smoothing=0.05, reprob=0.25, remode='pixel', recount=1, resplit=False, finetune='./checkpoints/labram-base.pth', model_key='model|module', model_prefix='', model_filter_name='gzp', init_scale=0.001, use_mean_pooling=False, disable_weight_decay_on_rel_pos_bias=False, nb_classes=5, output_dir='./checkpoints/speech/', log_dir='./log/speech/', device='cuda', seed=0, resume='', auto_resume=True, save_ckpt=True, start_epoch=0, eval=False, dist_eval=False, num_workers=10, pin_mem=True, world_size=1, local_rank=-1, dist_on_itp=False, dist_url='env://', enable_deepspeed=False, dataset='speech', data_path='./data_processed', rank=0, gpu=0, distributed=True, dist_backend='nccl')
Speech dataset - Train: 4500, Val: 750, Test: 750
Sampler_train = <torch.utils.data.distributed.DistributedSampler object at 0x797a2a8b3dd0>
Patch size = 200
Load ckpt from ./checkpoints/labram-base.pth
Load state_dict by model_key = model
Weights of NeuralTransformer not initialized from pretrained model: ['blocks.0.attn.q_bias', 'blocks.0.attn.v_bias', 'blocks.1.attn.q_bias', 'blocks.1.attn.v_bias', 'blocks.2.attn.q_bias', 'blocks.2.attn.v_bias', 'blocks.3.attn.q_bias', 'blocks.3.attn.v_bias', 'blocks.4.attn.q_bias', 'blocks.4.attn.v_bias', 'blocks.5.attn.q_bias', 'blocks.5.attn.v_bias', 'blocks.6.attn.q_bias', 'blocks.6.attn.v_bias', 'blocks.7.attn.q_bias', 'blocks.7.attn.v_bias', 'blocks.8.attn.q_bias', 'blocks.8.attn.v_bias', 'blocks.9.attn.q_bias', 'blocks.9.attn.v_bias', 'blocks.10.attn.q_bias', 'blocks.10.attn.v_bias', 'blocks.11.attn.q_bias', 'blocks.11.attn.v_bias']
Weights from pretrained model not used in NeuralTransformer: ['mask_token', 'pos_embed', 'lm_head.weight', 'lm_head.bias']
Pretrained weights loaded to original LaBraM model
Testing model output shape...
Test output shape: torch.Size([2, 192, 200])
Actual number of tokens: 192
Expected (num_electrodes * num_time_patches): 192
...
  (classifier): Sequential(
    (0): Rearrange('b n d -> b (n d)')
    (1): Linear(in_features=38400, out_features=800, bias=True)
    (2): ELU(alpha=1.0)
    (3): Dropout(p=0.1, inplace=False)
    (4): Linear(in_features=800, out_features=200, bias=True)
    (5): ELU(alpha=1.0)
    (6): Dropout(p=0.1, inplace=False)
    (7): Linear(in_features=200, out_features=5, bias=True)
  )
...
Epoch: [49]  [50/70]  eta: 0:00:01  lr: 0.000001  min_lr: 0.000001  loss: 0.4218 (0.4299)  class_acc: 0.9688 (0.9562)  loss_scale: 131072.0000 (131072.0000)  weight_decay: 0.0001 (0.0001)  grad_norm: 23.6771 (29.4157)  time: 0.0946  data: 0.0001  max mem: 5778
Epoch: [49]  [60/70]  eta: 0:00:00  lr: 0.000001  min_lr: 0.000001  loss: 0.4288 (0.4320)  class_acc: 0.9531 (0.9544)  loss_scale: 131072.0000 (131072.0000)  weight_decay: 0.0001 (0.0001)  grad_norm: 25.5949 (29.2178)  time: 0.0930  data: 0.0001  max mem: 5778
Epoch: [49]  [69/70]  eta: 0:00:00  lr: 0.000001  min_lr: 0.000001  loss: 0.4303 (0.4307)  class_acc: 0.9531 (0.9560)  loss_scale: 131072.0000 (131072.0000)  weight_decay: 0.0001 (0.0001)  grad_norm: 24.4372 (28.8118)  time: 0.0912  data: 0.0001  max mem: 5778
Epoch: [49] Total time: 0:00:06 (0.0961 s / it)
Averaged stats: lr: 0.000001  min_lr: 0.000001  loss: 0.4303 (0.4307)  class_acc: 0.9531 (0.9560)  loss_scale: 131072.0000 (131072.0000)  weight_decay: 0.0001 (0.0001)  grad_norm: 24.4372 (28.8118)
Val:  [0/8]  eta: 0:00:01  loss: 2.0136 (2.0136)  accuracy: 0.2708 (0.2708)  balanced_accuracy: 0.2627 (0.2627)  f1_weighted: 0.2737 (0.2737)  cohen_kappa: 0.0851 (0.0851)  time: 0.1772  data: 0.1381  max mem: 5778
Val:  [7/8]  eta: 0:00:00  loss: 1.9396 (1.9665)  accuracy: 0.3229 (0.3213)  balanced_accuracy: 0.3201 (0.3227)  f1_weighted: 0.3180 (0.3175)  cohen_kappa: 0.1464 (0.1498)  time: 0.0552  data: 0.0173  max mem: 5778
Val: Total time: 0:00:00 (0.0594 s / it)
* loss 1.966
Accuracy of the network on the 750 val EEG: 0.32%
Test:  [0/8]  eta: 0:00:01  loss: 1.7465 (1.7465)  accuracy: 0.3438 (0.3438)  balanced_accuracy: 0.3460 (0.3460)  f1_weighted: 0.3427 (0.3427)  cohen_kappa: 0.1788 (0.1788)  time: 0.1790  data: 0.1408  max mem: 5778
Test:  [7/8]  eta: 0:00:00  loss: 1.7103 (1.8060)  accuracy: 0.3125 (0.3360)  balanced_accuracy: 0.3097 (0.3317)  f1_weighted: 0.3109 (0.3333)  cohen_kappa: 0.1337 (0.1672)  time: 0.0555  data: 0.0176  max mem: 5778
Test: Total time: 0:00:00 (0.0590 s / it)
* loss 1.806
Accuracy of the network on the 750 test EEG: 0.34%
Max accuracy val: 0.32%, max accuracy test: 0.34%
Training time 0:07:07
